### PR TITLE
Fixed MaskedTextBoxAccessibleObject.Name if Owner.Mask is filled

### LINKED ISSUE #7938: For MaskedTextBox with Mask setting, the Name property in Inspect should be empty
### .NET version

8.0.100-alpha.1.22512.32

### Did it work in .NET Framework?

Yes

### Did it work in any of the earlier releases of .NET Core or .NET 5+?

Yes, cannot repro in previous .Net versions: 7.0/6.0

### Issue description

For MaskedTextBox with Mask setting, the Name property in Inspect should be empty. This issue is regressed from PR fixing: https://github.com/dotnet/winforms/pull/7741

**Actual result:**
image

**Expected result:**
image

### Steps to reproduce

**Repro steps:**
1.	Create a Winforms .Net 8.0 application
2.	Add several maskedTextBox controls
3.	Set them with different mask: PhoneNumeber, ShortDate
4.	Build and run application, use Inspect to check these controls' Name property

or use this attached sample application:
[WinFormsApp8.zip](https://github.com/dotnet/winforms/files/9774999/WinFormsApp8.zip)

### PR DESCRIPTION
**Pull request by NikitaSemenovAkvelon**
Fixes #7938

## Proposed changes

- Overriden `Name` property in `MaskedTextBoxAccessibleObject`.
- Added new unit tests. 

## Customer Impact

- The `Name` of `MaskedTextBoxAccessibleObject` is more clarify.

## Regression? 

- Yes

## Risk

- Minimal

### Before

Screenshot 2022-10-21 121439

### After

Screenshot 2022-10-21 120955

## Test methodology 

- Manually
- Unit testing 

## Accessibility testing  

- Inspect

 

## Test environment(s) 

.NET SDK:
 Version:   8.0.100-alpha.1.22512.5
 Commit:    1b80461e45

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.22621
 OS Platform: Windows

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/7983)

### CODE REVIEW

**Review by dmitrii-drobotov — APPROVED** [review]
Looks good, but I would add a comment to clarify the change

File: src/System.Windows.Forms/src/System/Windows/Forms/MaskedTextBox.MaskedTextBoxAccessibleObject.cs [hunk 1]
```diff
@@ -12,6 +12,14 @@ public MaskedTextBoxAccessibleObject(MaskedTextBox owner) : base(owner)
             {
             }
 
+            public override string? Name
+            {
+                get => string.IsNullOrEmpty((Owner as MaskedTextBox)!.Mask)
+                    ? base.Name
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/MaskedTextBox.MaskedTextBoxAccessibleObject.cs:18]: Do we have a test in WinformsControlsApp that covers MaskedTextBox? Could you please add one if not.

File: src/System.Windows.Forms/src/System/Windows/Forms/MaskedTextBox.MaskedTextBoxAccessibleObject.cs [hunk 2]
```diff
@@ -12,6 +12,14 @@ public MaskedTextBoxAccessibleObject(MaskedTextBox owner) : base(owner)
             {
             }
 
+            public override string? Name
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/MaskedTextBox.MaskedTextBoxAccessibleObject.cs:15]: I expected `UseTextForAccessibility` style to handle this case and not read the text box text.  https://github.com/dotnet/winforms/blob/55b05866b52ec11ce1eeedb89b6ab8d6b1ef9571/src/System.Windows.Forms/src/System/Windows/Forms/TextBoxBase.cs#L98-L102 

and
https://github.com/dotnet/winforms/blob/55b05866b52ec11ce1eeedb89b6ab8d6b1ef9571/src/System.Windows.Forms/src/System/Windows/Forms/Control.ControlAccessibleObject.cs#L311-L323

Do you see why this didn't work?

File: src/System.Windows.Forms/src/System/Windows/Forms/MaskedTextBox.MaskedTextBoxAccessibleObject.cs [hunk 3]
```diff
@@ -12,6 +12,14 @@ public MaskedTextBoxAccessibleObject(MaskedTextBox owner) : base(owner)
...
+            {
+                get => string.IsNullOrEmpty((Owner as MaskedTextBox)!.Mask)
+                    ? base.Name
+                    : base.Name ?? string.Empty;
```
[the remark this answers is not in the document]
Comment (RussKie) [reply src/System.Windows.Forms/src/System/Windows/Forms/MaskedTextBox.MaskedTextBoxAccessibleObject.cs:19]: I suggest against using `!` unless absolultely necessary, and this doesn't look like that case.
`Owner as MaskedTextBox` is superflous, see https://github.com/dotnet/winforms/issues/7974.

File: src/System.Windows.Forms/src/System/Windows/Forms/MaskedTextBox.MaskedTextBoxAccessibleObject.cs (hunk 1 above)
Comment (NikitaSemenovAkvelon) [reply src/System.Windows.Forms/src/System/Windows/Forms/MaskedTextBox.MaskedTextBoxAccessibleObject.cs:18]: CollectionEditors form contains MaskedTextBox. This form is on my screenshots above.

File: src/System.Windows.Forms/src/System/Windows/Forms/MaskedTextBox.MaskedTextBoxAccessibleObject.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (NikitaSemenovAkvelon) [reply src/System.Windows.Forms/src/System/Windows/Forms/MaskedTextBox.MaskedTextBoxAccessibleObject.cs:19]: Maybe the entire refactoring should be located in a single PR? Also, #7974 is not in progress yet.

File: src/System.Windows.Forms/src/System/Windows/Forms/MaskedTextBox.MaskedTextBoxAccessibleObject.cs (hunk 2 above)
Comment (NikitaSemenovAkvelon) [reply src/System.Windows.Forms/src/System/Windows/Forms/MaskedTextBox.MaskedTextBoxAccessibleObject.cs:15]: It's working. But in this case, it hasn't any Label.

- Without any changes, the `Name` gets a value from a previous `Label`:
Screenshot 2022-10-24 100512
- Set `UseTextForAccessibility` style to `true`. The name gets a value from the owner's `Text`:
Screenshot 2022-10-24 100620

The issue occurs because the `Name` property is changed to `Value` somewhere outside if we return null.

File: src/System.Windows.Forms/src/System/Windows/Forms/MaskedTextBox.MaskedTextBoxAccessibleObject.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (Tanya-Solyanik) [reply src/System.Windows.Forms/src/System/Windows/Forms/MaskedTextBox.MaskedTextBoxAccessibleObject.cs:19]: >Maybe the entire refactoring should be located in a single PR?

not necessarily, fill free to start here.

File: src/System.Windows.Forms/src/System/Windows/Forms/MaskedTextBox.MaskedTextBoxAccessibleObject.cs (hunk 2 above)
Comment (Tanya-Solyanik) [reply src/System.Windows.Forms/src/System/Windows/Forms/MaskedTextBox.MaskedTextBoxAccessibleObject.cs:15]: >The issue occurs because the Name property is changed to Value somewhere outside if we return null.

This is probably happening in the system IAccessible.

File: src/System.Windows.Forms/src/System/Windows/Forms/MaskedTextBox.MaskedTextBoxAccessibleObject.cs (hunk 3 above)
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/MaskedTextBox.MaskedTextBoxAccessibleObject.cs:19]: Please add a comment that we want to avoid the use of `Mask` as an AccessibleName

File: src/System.Windows.Forms/src/System/Windows/Forms/MaskedTextBox.MaskedTextBoxAccessibleObject.cs [hunk 4]
```diff
@@ -10,8 +10,21 @@ internal class MaskedTextBoxAccessibleObject : TextBoxBaseAccessibleObject
         {
             public MaskedTextBoxAccessibleObject(MaskedTextBox owner) : base(owner)
             {
+                Owner = owner;
             }
 
+            public override string? Name
+            {
+                get => string.IsNullOrEmpty((Owner as MaskedTextBox)!.Mask)
```
Comment (dmitrii-drobotov) [inline src/System.Windows.Forms/src/System/Windows/Forms/MaskedTextBox.MaskedTextBoxAccessibleObject.cs:18]: It should work now
```suggestion
                get => string.IsNullOrEmpty(Owner.Mask)
```

**Review by dmitrii-drobotov — DISMISSED** [review]

File: src/System.Windows.Forms/src/System/Windows/Forms/Control.ControlAccessibleObject.cs [hunk 5]
```diff
@@ -338,7 +338,7 @@ public IntPtr Handle
                 }
             }
 
-            public Control Owner { get; }
+            public virtual Control Owner { get; }
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/Control.ControlAccessibleObject.cs:341]: Sorry about stirring you into a wrong direction. This is a breaking change - 
```
Adding virtual to a member

While this change would often work without breaking too many scenarios because C# compiler tends to emit callvirt IL instructions to call non-virtual methods (callvirt performs a null check, while a normal call won't), we can't rely on it. C# is not the only language we target and the C# compiler increasingly tries to optimize callvirt to a normal call whenever the target method is non-virtual and the this is provably not null (such as a method accessed through the ?. null propagation operator). Making a method virtual would mean that consumer code would often end up calling it non-virtually.
```
https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/breaking-change-rules.md#source-and-binary-compatibility-changes

We could add a second property with the same content which would be internal virtual. But it's really best done in a separate PR.

Comment (RussKie) [reply src/System.Windows.Forms/src/System/Windows/Forms/Control.ControlAccessibleObject.cs:341]: Good call, @Tanya-Solyanik.

We should add a new internal property:
```cs
internal virtual Control OwnerInternal { get; }
```

**Review by dmitrii-drobotov — APPROVED** [review]

**Review by Tanya-Solyanik — APPROVED** [review]
